{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First I created a small CRM program, when given a nucleotide sequence, translates it into its protein counterpart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "subprocess.call('cat nuc2protein.crm', shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Talked with Jarrod, and we both came up with ideas about the direction of the project.\n",
    "\n",
    "###Initial Need\n",
    "\n",
    "Initially the idea was a fuzzy protein search. You input 'PXFARX{5}C' into the search and it gives you the location, promoters, etc. associated with derivatives of PFARQ. Basically we're trying to make a fuzzy search engine for [sequence motifs](https://en.wikipedia.org/wiki/Sequence_motif)\n",
    "\n",
    "[Promoter Region][PXFARX{5}CXXCH}[Region until Stop codon]\n",
    "\n",
    "N = A,G,C,T\n",
    "R = A/G\n",
    "Y = C/T\n",
    "\n",
    "PXFARX{5}CXXCH = CCN N{3} TTY GCN CGN|AGR N{15} TGY N{6} TGY CAY\n",
    "\n",
    "Someone already built a [Regex search engine for DNA](http://benchling.engineering/dna-regex-search/).\n",
    "\n",
    "###Derivatives of PFARQ\n",
    "\n",
    "We defined derivatives as amino acids that are similar in some way. For example, P (tryptophan) and Y (tyrosine) are both aromatic groups, so a derivative of PFARQ could be YFARQ for example.\n",
    "\n",
    "Other differentiators could be amino acid size, charge (+/-), acidic groups, basic groups, polar/nonpolar, purine vs. pyridine\n",
    "\n",
    "###How to find PFARQ derivatives\n",
    "As I was reading, I came across the concept of the Levenshtein edit distance. I mentioned it to Jarrod, and we thought a way we could merge the edit distance concept with the PFARQ derivatives.\n",
    "\n",
    "*Edit distance*: p*ho*to -> p*oh*to (1 operation, switch h and o)\n",
    "\n",
    "The genomics version would be with amino acid sequences. Say you have the amino acid F (phenylalanine). From to F -> [WY] (both W/Y are aromatic), so the edit distance weight would only be 0.5. From F->G (both are completely unrelated) so the edit distance weight would be 2.\n",
    "\n",
    "I came up with the idea of constructing a matrix of these weights between amino acids. \n",
    "\n",
    "   **P F A R Q**\n",
    "\n",
    "**P** .5 2 .7 ..\n",
    "\n",
    "**F**\n",
    "\n",
    "**A**\n",
    "..\n",
    "\n",
    "I then also thought about the idea of emergence, where maybe we input the weights in the matrix initially, but over time the most optimal weight arrangement would appear. \n",
    "\n",
    "Jarrod then suggested I can create synthetic sequences by changing the \"initial conditions\" of these weights and use a machine learning classifier/look through a database of known sequences to see if these synthetic sequences are legitimate.\n",
    "\n",
    "His inspiration coming from a [chaotic mapping to generate musical variations paper](http://www.eecs.qmul.ac.uk/~eniale/teaching/ise575/c/presentations/7-1-dabby_chuan.pdf) \n",
    "https://dspace.mit.edu/bitstream/handle/1721.1/27282/Dabby_Diana_PhD_1995.pdf?sequence=1\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "###Literature search\n",
    "\n",
    "Obviously before I try this, I want to see what other people have done.\n",
    "\n",
    "####Fuzzy string matching libraries/classifiers written in Python\n",
    "\n",
    " * Fuzzy string matching library in Python based off of Levenshtein distance (https://github.com/seatgeek/fuzzywuzzy)\n",
    "\n",
    " * [Fuzzy matching classifier from yhat](http://blog.yhat.com/posts/fuzzy-matching-with-yhat.html)\n",
    "\n",
    "\n",
    "####Motifs in DNA and their significance\n",
    "\n",
    "[A Guided Tour to Approximate String Matching] (http://www.ling.helsinki.fi/kit/2002s/ctl190net/materiaali/p31-navarro.pdf)\n",
    "\n",
    "With the rise of high-throughput sequencing techniques, DNA sequence data has become abundant. Recently, scientists have become more interested in searching for gene sequence motifs because they often are locations for sequence-specific binding sites or are involved in important RNA processes [6].. Motifs are short and recurring patterns in DNA (normally between 5 to 20 base-pairs long) that can be found similar genes in different organisms. Motifs can encode for structural motifs in protein sequences, which encode for the secondary structure of proteins (maybe explain more about importance of motifs) [2]. Many algorithms have been developed to identify motifs in DNA: Bailey and Elkan [3] developed the MEME algorithm based off of expectation maximization; Liu *et al* [4] used self-organizing neural networks to find motifs in DNA and protein sequences; other tools like CLUSTAL W [5] to first align multiple sequences and then identify motifs from conserved regions. \n",
    "\n",
    "List of other algorithms: [Quang, Daniel, and Xiaohui Xie. \"EXTREME: an online EM algorithm for motif discovery.\" Bioinformatics 30, no. 12 (2014): 1667-1673.], [Yao, Zizhen, Zasha Weinberg, and Walter L. Ruzzo. \"CMfinder—a covariance model based RNA motif finding algorithm.\" Bioinformatics 22, no. 4 (2006): 445-452.]\n",
    "\n",
    "Once a motif is identified, biologists are often interested in finding its occurences (need more info why, Jarrod any insight?) in other gene sequences (Talk about example with motifs being found in only certain types of organisms, in certain places, etc, any insight Jarrod?). DNA can be seen simply as long streams of text that adhere to an alphabet composed of A, C, G, and T, and thus computer algorithms that search for patterns in text can be used to find specified sequences like motifs. Unfortunately, exact searching of patterns is often futile because gene sequences may have small difference between them due to sequencing error, evolutionary differences between organisms, and mutations. Thus there is tremendous value in developing a *\"fuzzy (or approximate) search\"* that would allow a certain amount of error in the search [9]. \n",
    "\n",
    "One way to allow a certain amount of error in the search is via *edit distance*, most commonly exemplified by Levenshtein distance [10], a measure of the minimum number of operations needed to transform one string into another using only single-character insertions, deletions, or substitutions. In the evolution of DNA, certain kinds of point mutations via single-nucleotide deletions, insertions, and substitutions are more likely to occur than others [1]. The concept of edit distance can be adapted to consider this by assigning weights to certain editing operations; for example, assigning an increased penalty for mutating from a purine to a pyrimidine. This *weighted edit distance* allows one to select pattern matches that make biological sense. \n",
    "\n",
    "The next question becomes how can one assign weights to nucleotide insertions, deletions, and subsitutions to best reflect what the process of DNA mutation? For amino acid sequences, several substitution matrices that describe the probability of one amino acid mutating into another (say tyrosine into phenylalanine) based on calculated estimates from sequence alignment data, the most prominent examples being the PAM matrix (based on the rate of divergence between species) [12] and the BLOSUM matrix (based on chemical differences in aminoi acids) [13]. Substitution matrices have also been created to model nucleotide mutations, some based off transition and transversion effects [14].\n",
    "\n",
    "By using these probabilities from substitution matrices and adapting them into weight functions for edit operations in string search algorithms, one can design a fuzzy pattern search that can select matching patterns that make biological sense and reject others that do not. A previous paper that aimed to perform approximate DNA pattern searches [15] used a symmetric transvesion/transition weight function to account for the biological fact that purine/purine and pyrimidine/pyrimidine replacements (transitions) is more likely to occur than purine/pyrimidine replacements (transversions). However, this model ignores many factors that contribute to nucleotide mutation, such as observations of codon usage bias [CITATION NEEDED, any suggestions Jarrod?]. The same paper also transformed the PAM250 matrix [12] into a weight function to find fuzzy matches in protein sequences. While the PAM matrix does account for the probabilities of one amino acid mutatating into another, the correct matrix must be chosen as PAM matrices are optimized based on different degrees of evolutionary divergence [16]. In addition, the amino acid sequence may itself be beset with errors in the process of converting from raw DNA sequences to genes to amino acids [CITATION NEEDED, please help], so working in the nucleotide space is much preferred. \n",
    "\n",
    "Representing a motif found in amino acid space in terms of nucleotides can be done with a Position Weight Matrix (PWM) [17]. At each position of a PWM, the probability of each nucleotide is given (INSERT DIAGRAM/EXAMPLE HERE), an arrangement that is commonly visualized via sequence logos [7] in which motifs can be easily identified via visual inspection as seen in Figure 1. of [8]. PWMs are useful because amino acids can be encoded by several different codons and thus are best represented as probabilities of nucleotides. [18] presented methods for exact pattern matching using weighted sequences. However, given the fact that amino acids with similar proprties have been found to be able to be interchangable without affecting protein folding [CIATION NEEDED], there is still a need for an fuzzy pattern matching algorithm using weighted sequences in which amino acid motifs can be represented as nucleotides. Recently, researchers have theorized algorithms that define edit distances for weighted sequences, which if implemented, could possibly allow for an approximate matching of motif patterns [19].  \n",
    "\n",
    "A fuzzy search for motifs in the nucleotide space would enable scientists to answer some fundamental biological questions. Gene sequences that contain coding information for proteins are often more conserved than non-coding regions. These regions, which tend to contain DNA motifs, are often binding sites for proteins that regulate transcription [20]. In addition, fuzzy motif searches can help a biologist, who for example, wants to determine whether a motif (or a derivative of one) found in iron-oxidizing bacterial in saltwater can also be found the evolutionarily-divergent, yet iron-oxidizing bacterial in freshwater [cite Jarrod for the example!]\n",
    "\n",
    "####Methods \n",
    "\n",
    "The number of nucleotides in the PFARQ nucleotide sequences will be measured. Based on that number, nucleotide substitution matrices of the PFARQ sequence will be created via the methods described in [21], and from the matrices, PWMs can be created. In addition, in order to consider effects from amino acid chemical properties, PAM [12] and BLOSUM substitution matrices [13] will also be used to create PWMs as done in [15]. Meanwhile an implementation of the approximate matching algorithm for weighted sequences as described in [19] will either be found online or written in Python/C++, after which the algorithm will be used to search for protein motifs represented in the nucleotide space as PWMs. As a benchmark, the PWM match search program, MOODS [22] will also used to search for motifs.   \n",
    "\n",
    "The example protein motif that will be used is called the PFARQ motif, which using IUPAC notation can be described in the amino acid space as \"PXFARX{5}CX{2}CH\". The nucleotide-space equivalent of the motif can be described as \"CCN N{3} TTY GCN CGN|AGR N{15} TGY N{6} TGY CAY\". 125 amino acid sequences containing the PFARQ motif from NCBI and JGI in the fasta format and their corresponding nucleotide sequences will be used to validate the efficacy of the fuzzy motif search. In addition, sequences containing PFARQ derivatives will either be artificially generated or found and manually labeled for use as a gold standard in terms of PFARQ sequences. Another option is to test the fuzzy motif search algorithm on sequences that contain known motifs in the amino acid space; a script to convert the amino acid sequences to nucleotides can also be written if necessary to generate gold standard data to validate the algorithm on.\n",
    "\n",
    "The main challenge falls in the evaluation of any algorithm's effectiveness. If given a labeled set of motif sequences and their derivatives, the fuzzy search algorithms can attempt to search for the motifs and generate scoring metrics such as precision and recall. This process can be repeated many times with the only difference between runs being the selection of the substitution matrix to use (BLOSUM, PAM, methods described in [21]). An ROC curve could then be generated from the different runs. This evaluation of accuracy the nucleotide substitution matrices in [21] using actual biologically sequences in itself is very valuable for research; one disadvantage in the methodology of [21] is the use of random rate matrices instead of those based off of biological effects. Optimization algorithms (perhaps neural networks or a specified training method) could then possibly be applied to improve precision and accuracy of the fuzzy motif search algorithm. The limiting factor falls in the amount of labeled sequences of motifs that also account for their derivatives.\n",
    "\n",
    "\n",
    "####Reference\n",
    "\n",
    "1. Freese, Ernst. \"The difference between spontaneous and base-analogue induced mutations of phage T4.\" Proceedings of the National Academy of Sciences 45, no. 4 (1959): 622-633.\n",
    "\n",
    "2. Das, Modan K., and Ho-Kwok Dai. \"A survey of DNA motif finding algorithms.\" BMC bioinformatics 8, no. Suppl 7 (2007): S21.\n",
    "\n",
    "3. Bailey, Timothy L., and Charles Elkan. \"Unsupervised learning of multiple motifs in biopolymers using expectation maximization.\" Machine learning 21, no. 1-2 (1995): 51-80.\n",
    "\n",
    "4. Liu, Derong, B. DasGupta, and Huaguang Zhang. \"Motif discoveries in unaligned molecular sequences using self-organizing neural networks.\" Neural Networks, IEEE Transactions on 17, no. 4 (2006): 919-928.\n",
    "\n",
    "5. Thompson, Julie D., Desmond G. Higgins, and Toby J. Gibson. \"CLUSTAL W: improving the sensitivity of progressive multiple sequence alignment through sequence weighting, position-specific gap penalties and weight matrix choice.\" Nucleic acids research 22, no. 22 (1994): 4673-4680.\n",
    "\n",
    "6. D'haeseleer, Patrik. \"What are DNA sequence motifs?.\" Nature biotechnology 24, no. 4 (2006): 423-425.\n",
    "\n",
    "7. Schneider, Thomas D., and R. Michael Stephens. \"Sequence logos: a new way to display consensus sequences.\" Nucleic acids research 18, no. 20 (1990): 6097-6100.\n",
    "\n",
    "8. D'haeseleer, Patrik. \"What are DNA sequence motifs?.\" Nature biotechnology 24, no. 4 (2006): 423-425.\n",
    "\n",
    "9. Navarro, Gonzalo. \"A guided tour to approximate string matching.\" ACM computing surveys (CSUR) 33, no. 1 (2001): 31-88.\n",
    "\n",
    "10. Levenshtein, Vladimir I. \"Binary codes capable of correcting deletions, insertions, and reversals.\" In Soviet physics doklady, vol. 10, no. 8, pp. 707-710. 1966.\n",
    "\n",
    "11. Behura, Susanta K., and David W. Severson. \"Codon usage bias: causative factors, quantification methods and genome‐wide patterns: with emphasis on insect genomes.\" Biological Reviews 88, no. 1 (2013): 49-61.\n",
    "\n",
    "12. Dayhoff, Margaret O., and Robert M. Schwartz. \"A model of evolutionary change in proteins.\" In In Atlas of protein sequence and structure. 1978.\n",
    "\n",
    "13. Henikoff, Steven, and Jorja G. Henikoff. \"Amino acid substitution matrices from protein blocks.\" Proceedings of the National Academy of Sciences 89, no. 22 (1992): 10915-10919.\n",
    "\n",
    "14. Li, Wen-Hsiung. \"Unbiased estimation of the rates of synonymous and nonsynonymous substitution.\" Journal of molecular evolution 36, no. 1 (1993): 96-99.\n",
    "\n",
    "15. Kurtz, Stefan. \"Approximate string searching under weighted edit distance.\" In Proc. WSP, vol. 96, pp. 156-170. 1996.\n",
    "\n",
    "16. Altschul, Stephen F., John C. Wootton, E. Michael Gertz, Richa Agarwala, Aleksandr Morgulis, Alejandro A. Schäffer, and Yi‐Kuo Yu. \"Protein database searches using compositionally adjusted substitution matrices.\" Febs Journal 272, no. 20 (2005): 5101-5109.\n",
    "\n",
    "17. Thompson, Julie D., Desmond G. Higgins, and Toby J. Gibson. \"CLUSTAL W: improving the sensitivity of progressive multiple sequence alignment through sequence weighting, position-specific gap penalties and weight matrix choice.\" Nucleic acids research 22, no. 22 (1994): 4673-4680.\n",
    "\n",
    "18. Christodoulakis, Manolis, Costas S. Iliopoulos, Laurent Mouchard, and Kostas Tsichlas. \"Pattern matching on weighted sequences.\" In Proceedings of the Algorithms and Computational Methods for Biochemical and Evolutionary Networks (CompBioNets), pp. 17-30. 2004.\n",
    "\n",
    "19. Amir, Amihood, Costas Iliopoulos, Oren Kapah, and Ely Porat. \"Approximate matching in weighted sequences.\" In Combinatorial Pattern Matching, pp. 365-376. Springer Berlin Heidelberg, 2006.\n",
    "\n",
    "20. Stormo, Gary D. \"DNA binding sites: representation and discovery.\" Bioinformatics 16, no. 1 (2000): 16-23.\n",
    "\n",
    "21. Oscamou, Maribeth, Daniel McDonald, Von Bing Yap, Gavin A. Huttley, Manuel E. Lladser, and Rob Knight. \"Comparison of methods for estimating the nucleotide substitution matrix.\" BMC bioinformatics 9, no. 1 (2008): 511.\n",
    "\n",
    "\n",
    "\n",
    "###Another interesting topic: Substitution matrices for amino acids/nucleotides \n",
    "\n",
    "**Answers the question of \"what should my edit weights be?\"**\n",
    "\n",
    "PAM1 means probability of each amino acid changing to another is ~1% and probability of not changing is ~99% (doesn't seem legit...). [This link explains the PAM matrix really well](http://www.math.lsa.umich.edu/~dburns/seqalmit2.pdf). So the PAM250 Matrix is a table of the expected change in amino acids?? Is there a way to consolidate this model with nucleotide changes, since this model appears to ignore that. PAM250 matrix means 250% expected change. \n",
    "\n",
    "Dayhoff et al. (1978). A model of evolutionary change in proteins. In Atlas of Protein\n",
    "Sequence and Structure, vol. 5, suppl. 3, 345–352. National Biomedical Research\n",
    "Foundation, Silver Spring, MD, 1978\n",
    "\n",
    "Expected Similarity: PAM120-40%, PAM80-50%, PAM60-60%, PAM250-15030% similarity. You have to run all of them and check which gives highest ungapped alignment score.\n",
    "\n",
    "*Derivation*: maximum parsimony (emphasizes least number of substitutions), relative mutability (ratio of total # of times amino acid has changed vs. # of occurences), **relative mutability** based on chemical properties, effective frequency (consider difference in variability of primary protein structure), **mutational probability matrix** (probability of amino acid being substituted over a given unit of evolutionary time), log-odd (used most in practice)\n",
    "\n",
    "tl;dr: [summary of types of substitution matrices for amino acids (PAM, BLOSUM, etc.)](http://homepages.ulb.ac.be/~dgonze/TEACHING/pam_blosum.pdf))\n",
    "\n",
    "####Substitution matrices for protein sequences\n",
    "\n",
    "**PAM Matrix is based on rate of divergence between sequences** (. PAMn = PAM1^n (gives the probability to observe the changes i->j per 100xn mutations). *J. van Helden (log-odd PAM250)*\n",
    "\n",
    "Schwartz RM, Dayhoff M. Matrices for detecting distant relationships. In: Dayhoff M, editor. Atlas of Protein Sequence and Structure. supplement 3. volume 5. National Biomedical Research Foundation; Silver Spring, MD: 1978. pp. 353–358.\n",
    "\n",
    "**BLOSUM Substitution matrix - designed to find conserved regions in proteins**, not based on evolutionary distances, based on protein chemical similarities (considers hydrophobia, aromaticity, polarity, basicity, acidity)\n",
    "\n",
    " *Henikoff S and Henikoff JG (1992). Amino acid substitution matrices from\n",
    "protein blocks. PNAS 89:10915-10919.* \n",
    "\n",
    "Improved BLOSUM62 (MP Styczynski, KL Jensen, I Rigoutsos, G Stephanopoulos\n",
    "Nat. Biotech. 26: 274–275, 2008)\n",
    "\n",
    "**GONNET Substitution matrix** - based on exhaustive sequence alignment analysis(Gonnet, Cohe\n",
    "\n",
    "####Substitution matrices for nucleotides\n",
    "\n",
    "Considering transition (purine to purine, pyrimidine to pyrimidine) and transversion effects (purine -> pyrimidine and vice versa)\n",
    "\n",
    "Li, Wen-Hsiung. \"Unbiased estimation of the rates of synonymous and nonsynonymous substitution.\" Journal of molecular evolution 36, no. 1 (1993): 96-99.\n",
    "\n",
    "See: http://homepages.ulb.ac.be/~dgonze/TEACHING/pam_blosum.pdf \n",
    "\n",
    "Models of DNA evolution: https://en.wikipedia.org/wiki/Models_of_DNA_evolution#Continuous-time_Markov_chains \n",
    "\n",
    "####Idea: use the probability distribution found in PAM matrices as a basis for amino acid mutations in project. use PWM as scoring metric. MACHINE LEARN THESE TABLES\n",
    "\n",
    "####Which matrix do we choose?\n",
    "\n",
    "[Selecting the Right Similarity-Scoring Matrix - Pearson, 2013](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3848038/): use shallow scoring matrices (VTML10-80) with short protein domains or with organisms that may have recently diverged, use deep scoring matrices (BLOSUM62) for longer alignments with 20-30% identity\n",
    "How to choose appropriate PAM matrix? PAM120 most appropriate for database searches, PAM200 most appropriate for comparing two specific proteins with suspected homology (Altschul SF, 1991)\n",
    "\n",
    "**Comparison of Methods for estimating the nucleotide substitution matrix - Oscamou, 2008** - suggests the use of Gojobori for >600 nucleotides, Goldman for < 600 nucleotides, and Barry/Hartigan for >2000 nucleotides. However, this study uses random rate matrices and not biologically inspired ones. Our study will revisit this via machine learning\n",
    "\n",
    "####Improving the substitution matrix\n",
    "\n",
    "The Levenshtein distance So in terms of a \"distance function\" that would be most appropriate, then the Levenshtein distance would be most appropriate. \n",
    "\n",
    "This PDF (https://web.stanford.edu/class/cs124/lec/med.pdf) has a pretty good explanation of the \"weighted minimum edit distance\" which is what I'm interested in, but so far it's in the realm of sequence alignment.\n",
    "\n",
    "*Citation Bank*:  Needleman and Wunsch [1970], Waterman [1995], Wagner-Fischer algorithm, \n",
    "[Gusfield 1997 (p.224) on weighted edit distance](http://s3.amazonaws.com/academia.edu.documents/31053376/Algorithms_on_String_Trees_and_Sequences.pdf?AWSAccessKeyId=AKIAJ56TQJRTWSMTNPEA&Expires=1466020763&Signature=vf5TMiTSz4AhH%2BrbUI%2FF%2FgboGvw%3D&response-content-disposition=inline%3B%20filename%3DAlgorithms_on_strings_trees_and_sequence.pdf), \n",
    "\n",
    "Basically in my searches (in which I discover that [Ben Langmead is a legend](http://www.langmead-lab.org/teaching-materials/)), the minimum edit distance concept seems to be mostly applied to alignment type things. Let's see if I can find papers that show it being applied in a PFARQ-like manner...\n",
    "\n",
    "Distance d(A,B): minimum number of operations needed to transform genome A into genome B\n",
    "\n",
    "\n",
    "\n",
    "###How do I represent my weighted edit distances? AKA What is a PWM (position weight matrix)?\n",
    "\n",
    "Has 4 rows (if doing nucleotide comparison) or 20 rows for amino acids. First, create a position frequency matrix (PFM) by counting occurences of each nucleotide at each position. Then create a position probability matrix (PPM) by dividing that former nucleotide count by # of sequences. Then you figure how frequently each letter appears in the dataset; assuming each letter appears equally then probability distribution (bK) would be 0.25 for nucleotides and 0.05 for amino acids. So you do some ln(m/bK) to transform it.\n",
    "\n",
    "When you add the relevant values at each position in the PWM, you get a \"sequence score\" which measures how different the sequence is from a random one. Score of 0 indicates same probability of being both a random/non-random site. Greater than 0 means more likely to be non-random, etc.\n",
    "\n",
    "\n",
    "\n",
    "The wikipedia article on [sequence motifs](https://en.wikipedia.org/wiki/Sequence_motif) led me to a bunch more papers (like this related but not quite relevant one: [Finding Tandem sequences (sequences that repeat one after another) algorithm](http://bioinformatics.oxfordjournals.org/content/23/2/e30.full))\n",
    "\n",
    "###Motif search papers\n",
    "\n",
    "* MaMF [A deterministic motif finding algorithm with application to the human genome, Hon and Jain (2006)](http://bioinformatics.oxfordjournals.org/content/22/9/1047.full): given a set of N promoters and motif width (w), MaMF can find high scoring motifs across different promoters. Their scoring function is based on motif conservation, uniqueness of motif relative to the genome (seeing how many motifs and approximate cousins live in the same genome?). They align two motifs using PWMs.\n",
    "\n",
    "* [EXTREME: an online EM algorithm for motif discovery](http://bioinformatics.oxfordjournals.org/content/30/12/1667.full.pdf+html) (it's more advanced cousin using neural networks http://nar.oxfordjournals.org/content/early/2016/04/15/nar.gkw226.full.pdf+html)\n",
    "\n",
    "\n",
    "####References\n",
    "* [Use of perceptron algorithm to distinguish translational initiation sites in E.Coli (the OG paper that spawned the PWM) by Stormo, 1982](http://nar.oxfordjournals.org/content/10/9/2997.full.pdf+html)\n",
    "\n",
    "* [Prior knowledge in finding Motifs using MEME](http://www.aaai.org/Papers/ISMB/1995/ISMB95-003.p\n",
    "* (J. Korhonen, P. Martinmäki, C. Pizzi, P. Rastas and E. Ukkonen. MOODS: fast search for position weight matrix matches in DNA sequences. Bioinformatics 25(23), pages 3181-3182. (2009)) \n",
    "\n",
    "###Pattern Matching on Weighted Sequences AKA fuzzy matching for protein sequences\n",
    "* **[Pattern Matching on Weighted Sequences (Christodoulakis, 2004)](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.63.604&rep=rep1&type=pdf), Approximate Matching in Weighted Sequences - Amir 2006.** - I think this is the jackpot. Basically PXFARX{5}CXXCH devolves into CCN N{3} TTY GCN CGN|AGR N{15} TGY N{6} TGY CAY. *This nucleotide sequence can be represented by a Position Weight Matrix, of which we can then apply Pattern Matching* Probably need to somehow bring in BLOSUM65 matrices into this...\n",
    "\n",
    "\n",
    "\n",
    "##side idea: machine learning actual gene sequences, creating synthetic ones, blasting them to get scores, building 2d protein structures, assess stability, retrain\n",
    "\n",
    "###Reading list for that:\n",
    "Enhanced Regulatory Sequence Prediction Using Gapped k-mer Features\n",
    "\n",
    "\n",
    "###DCJ (Double Cut Join) operations - model for genome rearrangement to define edit distance based on order and orientation (inversions and transversions)\n",
    "Model for edit distance based on gene order and orientation rather than nucleotide sequences. (lmao not what I want I don't think...). Basically Bergeron, et al defined DCJ distance as some graph theory thing. \"The genomic distance problem in the Hannenhalli-Pevzner theory is the following: Given two genomes whose chromosomes are linear, calculate the minimum number of inversions and translocations that transform one genome into the other.\" - so this is the basis for synteny algorithms...\n",
    "\n",
    "Friedberg R, Darling AE, Yancopoulos S (2008). \"Genome rearrangement by the double cut and join operation.\". Methods Mol Bio 285 (452): 385–416. doi:10.1007/978-1-60327-159-2_18. PMID 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
